runtime/lua/vim/treesitter/query.lua

```diff
@@ -92,6 +92,77 @@ local function read_query_files(filenames)
   return table.concat(contents, '\n')
 end
 
+local QueryMatch = {}
+
+-- Creates a query match.
+-- Contains metadata for the whole match or each individual capture in the match.
+-- Match nodes can be accessed by capture id.
+-- All nodes can be iterated over using the `iter()` method.
+function QueryMatch.new(match)
```

Comment:
hmm nothing about this class seems particular to queries or to matches. I have the suspicion that you could just pass in /return both the `match` and `metadata` tables instead of this "object" (and thus make `metadata` optional)

Reply:
I don't understand quite what you mean exactly. Could you clarify? This object holds match nodes and match metadata for a node or the whole match from a query result. It also adds an api for setting/getting metadata on the match table.

Reply:
> This object holds match nodes and match metadata for a node or the whole match from a query result. It also adds an api for setting/getting metadata on the match table.

This what I meant. you could just pass/return these two values, the match (as it was before), and the metadata table.

Reply:
If we go that route we need some way to get the metadata out to the consumer. We would have to emit it from the iterator as a third return value instead of having the data attached to the match.

```lua
for pattern, match, metadata in query:iter_matches(...) do
end
```

This would also require directives to take an additional metadata table to write to instead of using the api on the match. If these changes are fine, I will implement them.

Reply:
@vigoux Should we separate the metadata from the match that I mentioned above?

Reply:
I think is better as then a plugin that don't care about metadata (as the plugin does its own filtering from the matches) can just get matches as a simple table. The alternative would be a table with both `nodes` and `metadata` attribute but secret metatable with secret upvales is an anti-pattern. lua objects do not have "slots" and we should not pretend they do.